### PR TITLE
Make QualifierDefault extendable

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
Make `QualifierDefault` extendable, so that CF Inference could extend this base class and have an `InferenceQualifierDefault` for applying defaults in inference mode.

Original PR review in `opprop`: https://github.com/opprop/checker-framework/pull/61

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -497,8 +497,8 @@ public AnnotatedDeclaredType fromNewClass(NewClassTree newClassTree) {
     // Both methods should have some functionality merged into a single location.
     // See Issue 683
     // https://github.com/typetools/checker-framework/issues/683
-    protected final QualifierDefaults createQualifierDefaults() {
-        QualifierDefaults defs = new QualifierDefaults(elements, this);
+    protected final QualifierDefaults createAndInitQualifierDefaults() {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Can you also update the javadoc to reflect the name change.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -508,6 +508,10 @@ protected final QualifierDefaults createQualifierDefaults() {
         return defs;
     }
 
+    protected QualifierDefaults createQualifierDefaults() {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Add some javadoc, similar to what `createAndInitQualifierDefaults` has.

**Comment by wmdietl** [comment]
@CharlesZ-Chen Travis seems unhappy with merge problems: https://travis-ci.org/typetools/checker-framework/jobs/339126999
This is strange, as you haven't touched many of the files it complains about.
I've restarted the build once, which didn't help.

Can you update from master to see whether that fixes the issue?

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -484,10 +484,12 @@ public AnnotatedDeclaredType fromNewClass(NewClassTree newClassTree) {
     }
 
     /**
-     * Create {@link QualifierDefaults} which handles checker specified defaults. Subclasses should
-     * override {@link GenericAnnotatedTypeFactory#addCheckedCodeDefaults(QualifierDefaults defs)}
-     * or {@link GenericAnnotatedTypeFactory#addUncheckedCodeDefaults(QualifierDefaults defs)} to
-     * add more defaults or use different defaults.
+     * Create {@link QualifierDefaults} which handles checker specified defaults, and initialize the
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: How about simply

```
     * Create {@link QualifierDefaults} which handles checker specified defaults, and initialize the
     * created {@link QualifierDefaults}. Subclasses should override {@link
...
```

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -508,6 +510,14 @@ protected final QualifierDefaults createQualifierDefaults() {
         return defs;
     }
 
+    /**
+     * Create {@link QualifierDefaults} which handles checker specified defaults. Sub-classes
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]:
```
+     * Create {@link QualifierDefaults} which handles checker specified defaults. Subclasses
+     * should override this method to provide a different {@code QualifierDefault} implementation.
```

**Comment by CharlesZ-Chen** [comment]
@wmdietl I and Sadaf both meet this problem when prepare a PR against typetools --- even if we didn't touch other files, as long as our branch is behind typetools/master, travis will complains about merge conflicts....

I've update my branch to even with the typetools/master, this should fix the travis failure.

**Comment by mernst** [comment]
I see the merge problem once in a while too -- but not every time, and I don't know how to reproduce it.
(I presume that when you did the merge from upstream, it was clean, without the need for any human intervention at all.  Is that true?)
The logic in `.travis.yml` for merging should perhaps be re-examined and possibly improved.

**Comment by CharlesZ-Chen** [comment]
> (I presume that when you did the merge from upstream, it was clean, without the need for any human intervention at all. Is that true?)

Yes, that is true. opprop/master is always clean and has zero divergence from typetools/master. 
So it is strange that travis complains the merge conflict when preparing a PR from opprop to typetools....

**Pull request merged**